allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+            span candidates.
+        Returns
+        -------
+        attended_text_embeddings : ``torch.FloatTensor``
+            A tensor of shape (batch_size, num_spans, embedding_dim) - the result of
+            applying attention over all words within each candidate span.
+        """
+        # Shape: (1, 1, max_span_width)
+        max_span_range_indices = util.get_range_vector(self._max_span_width,
+                                                       text_embeddings.is_cuda).view(1, 1, -1)
+
+        # Shape: (batch_size, num_spans, max_span_width)
+        # This is a broadcasted comparison - for each span we are considering,
+        # we are creating a range vector of size max_span_width, but masking values
+        # which are greater than the actual length of the span.
+        span_mask = (max_span_range_indices <= span_width).float()
+        raw_span_indices = span_ends - max_span_range_indices
+        # We also don't want to include span indices which are less than zero,
+        # which happens because some spans near the beginning of the document
+        # are of a smaller width than max_span_width, so we add this to the mask here.
+        span_mask = span_mask * (raw_span_indices >= 0).float()
+        # Spans
+        span_indices = F.relu(raw_span_indices.float()).long()
+
+        # Shape: (batch_size * num_spans * max_span_width)
+        flat_span_indices = util.flatten_and_batch_shift_indices(span_indices, text_embeddings.size(1))
+
+        # Shape: (batch_size, num_spans, max_span_width, embedding_dim)
+        span_text_embeddings = util.batched_index_select(text_embeddings, span_indices, flat_span_indices)
+
+        # Shape: (batch_size, num_spans, max_span_width)
+        span_head_scores = util.batched_index_select(head_scores, span_indices, flat_span_indices).squeeze(-1)
+
+        # Shape: (batch_size, num_spans, max_span_width)
+        span_head_weights = util.last_dim_softmax(span_head_scores, span_mask)
+
+        # Do a weighted sum of the embedded spans with
+        # respect to the normalised head score distributions.
+        # Shape: (batch_size, num_spans, embedding_dim)
+        attended_text_embeddings = util.weighted_sum(span_text_embeddings, span_head_weights)
+
+        return attended_text_embeddings
+
+    def _compute_span_representations(self,
+                                      text_embeddings: torch.FloatTensor,
+                                      text_mask: torch.FloatTensor,
+                                      span_starts: torch.IntTensor,
+                                      span_ends: torch.IntTensor) -> torch.FloatTensor:
+        """
+        Computes an embedded representation of every candidate span. This is a concatenation
+        of the contextualized endpoints of the span, an embedded representation of the width of
+        the span and the result of applying attention to the words in the span.
```

Comment:
`s/the result of applying attention to the words in the span/a representation of the span's predicted head/`, or something like that.